src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -486,5 +523,16 @@ private static Pipe CreateDataPipe(MemoryPool<byte> pool)
                 useSynchronizationContext: false,
                 minimumSegmentSize: pool.GetMinimumSegmentSize()
             ));
+
+        public void Dispose()
```

Comment:
Lets get rid of the disposal altogether.

Reply:
So the loop in the background task will be `do { } while (true);`

I'll add a comment that it is never exited and will "exit" when the state machine is garbage collected along with Http2OutputProducer.

Reply:
I'm curious whether the state machines will be rooted by Task.s_currentActiveTasks when run under a debugger. That's a problem I noticed when writing the CriticalErrorLoggedIfApplicationDoesntComplete test.

https://github.com/dotnet/aspnetcore/blob/4a7bf75e89e3dfe7264a641ed041f50788d28d20/src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpConnectionManagerTests.cs#L24-L29

@davidfowl thinks this won't be an issue since ProcessDataWrites is ValueTask-returning, but it's probably worth double checking. The improved simplicity may not be worth it if it leads to a memory leak with a debugger attached.

Reply:
I think it will be a problem

Leak from IAsyncEnumerable which has `ValueTask<T> MoveNextAsync()`

https://github.com/dotnet/roslyn/issues/39321

Reply:
Toub mentioned making `s_currentActiveTasks` a weak table. It is currently still a dictionary: https://github.com/dotnet/runtime/search?q=s_currentActiveTasks&unscoped_q=s_currentActiveTasks

Reply:
```cs
static void Main(string[] args)
{
    for (int i = 0; i < 1000; i++)
    {
        _ = DoAThingValueTask();
        _ = DoAThingTask();
    }
    var t = Task.CompletedTask;
    Console.WriteLine("Hello World!");
    Console.ReadKey();
}

private static async ValueTask<ReadResult> DoAThingValueTask()
{
    Pipe p = new Pipe();
    return await p.Reader.ReadAsync();
}

private static async Task<ReadResult> DoAThingTask()
{
    Pipe p = new Pipe();
    return await p.Reader.ReadAsync();
}
```

Both are tracked:
![image](https://user-images.githubusercontent.com/303201/76268971-d25b2980-62d4-11ea-9aeb-791a6ce43213.png)

Reply:
I agree with @halter73 that avoiding a little complexity around Dispose isn't worth memory leaks when built as debug.